src/jbi/router.py

```diff
@@ -1,34 +1,87 @@
 """
 Router dedicated to Jira Bugzilla Integration APIs
 """
+import importlib
 import logging
-from typing import Optional
+import traceback
+from types import ModuleType
+from typing import Dict, List, Optional
 
-from fastapi import APIRouter, Depends, Request
-from fastapi.responses import HTMLResponse
+from fastapi import APIRouter, Body, Depends, Request
+from fastapi.responses import HTMLResponse, JSONResponse
 from fastapi.templating import Jinja2Templates
 
 from src.app import environment
 from src.jbi import configuration
+from src.jbi.bugzilla_objects import BugzillaBug, BugzillaWebhookRequest
+from src.jbi.service import get_bugzilla
 
 templates = Jinja2Templates(directory="src/templates")
 
 api_router = APIRouter(tags=["JBI"])
 
-jbi_logger = logging.getLogger("src.jbi")
+jbi_logger = logging.getLogger("src.jbi.router")
 
 
-def execute_request(request, settings):
+class ValidationError(Exception):
+    """Error throw when requests are invalid and ignored"""
+
+
+def extract_current_action(  # pylint: disable=inconsistent-return-statements
+    bug_obj: BugzillaBug, action_map, settings
+):
+    """Find first matching action from bug's whiteboard list"""
+    potential_configuration_tags: List[
+        str
+    ] = bug_obj.get_potential_whiteboard_config_list()
+    if not potential_configuration_tags:
+        return {}
+    for tag in potential_configuration_tags:
+        value = action_map.get(tag.lower())
+        if value:
+            return value
+
+
+def execute_request(request: BugzillaWebhookRequest, action_map, settings):
     """Execute action"""
+    try:
+        if not request.bug:
+            raise ValidationError("no bug data received")
+
+        is_private_bug = request.bug.is_private
+        if is_private_bug:
+            raise ValidationError("private bugs are not valid")
+
+        bugzilla_client = get_bugzilla()
+        current_bug_info = bugzilla_client.getbug(request.bug.id)
+        bug_obj = BugzillaBug.parse_obj(current_bug_info.__dict__)
+        current_action = extract_current_action(bug_obj, action_map, settings)
+        if not current_action:
+            raise ValidationError("bug does not have matching config")
+
+        action_module: ModuleType = importlib.import_module(current_action["action"])
+        if not action_module:
+            raise ValidationError("action not found")
+
+        callable_action = action_module.init(  # type: ignore
+            **current_action["parameters"]
+        )
+        return callable_action(payload=request)
+    except ValidationError as exception:
+        return JSONResponse(content={"error": str(exception)}, status_code=201)
```

Comment:
Should we use some `4xx` error here?

Reply:
The reason we don't use 4xx here is because these Errors are more of a Filtration component than an "error"
when these "invalid" requests are received we want to just move on and get the next message in the bugzilla webhook queue--a 4xx results in bugzilla retrying the request, since we want it "dropped" from the queue we need to give it a "success" since it was received and "processed".

_Perhaps using error for this is not appropriate..._

Reply:
I'm also doubtful about that. 

Are you sure Bugzilla retries on 4XX? 

Retrying by sending the same request twice would make sense for a 5XX response. But for a 4XX response, the client has to retry after having changed the request. 

In any case, 201 is not the appropriate code, since it means «successfully created».

Also, if an action raises an `ActionError` (because JIRA is down or whatever), then we want it to be retried by Bugzilla.

Reply:
```Errors

If the response is not 200, the Bugzilla system will retry 4 attempts, waiting 5s, 10s, 15s and 20s between each attempt. If none of those attempts is successful, the system will continue trying every 15 minutes.

If a message is stuck without a successful attempt, the next messages that trigger the webhook will be stored in a queue in the order that were triggered and will be delivered in that order when the first message in the queue is delivered.```

https://bugzilla.mozilla.org/page.cgi?id=webhooks.html

Reply:
Perhaps [202](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/202) then?